include/GIL/InstBase.hpp

```diff
@@ -0,0 +1,235 @@
+        assert(kind == OperandKind::LiteralIntKind && "Operand is not a literal integer");
+        return data.literalInt;
+    }
+    llvm::APFloat getLiteralFloat() {
+        assert(kind == OperandKind::LiteralFloatKind && "Operand is not a literal float");
+        return data.literalFloat;
+    }
+    llvm::StringRef getLiteralString() {
+        assert(kind == OperandKind::LiteralStringKind && "Operand is not a literal string");
+        return data.literalString;
+    }
+    // Function *getSymbol() {
+    //     assert(kind == OperandKind::SymbolKind && "Operand is not a symbol");
+    //     return data.symbol;
+    // }
+    // Type getType() {
+    //     assert(kind == OperandKind::TypeKind && "Operand is not a type");
+    //     return data.type;
+    // }
+    // Member getMember() {
+    //     assert(kind == OperandKind::MemberKind && "Operand is not a member");
+    //     return data.member;
+    // }
+    BasicBlock *getLabel() {
+        assert(kind == OperandKind::LabelKind && "Operand is not a label");
+        return data.label;
+    }
+};
+
+/**
+ * @class InstBase
+ * @brief Represents the base class for instructions in the GIL (Glu Intermediate Language).
+ *
+ * This class provides the basic interface for all instructions, including methods to
+ * get the kind of instruction, its operands, and its results. It also maintains a reference
+ * to the basic block that contains this instruction.
+ *
+ * @note This is an abstract class and cannot be instantiated directly.
+ */
+class InstBase {
+    /// The basic block that contains this instruction.
+    BasicBlock *parent = nullptr;
+    friend class BasicBlock; // Allow BasicBlock to set itself as the parent when added.
+public:
+    InstBase() {}
+    virtual ~InstBase() = default;
+    InstBase(const InstBase &) = delete;
+    InstBase &operator=(const InstBase &) = delete;
+    InstBase(InstBase &&) = delete;
+    InstBase &operator=(InstBase &&) = delete;
+
+    // These methods must be implemented by derived classes.
+
+    /// Returns the kind of this instruction, which is a value from the InstKind enumeration.
+    virtual InstKind getKind() = 0;
+    /// Returns the number of results produced by this instruction. For terminator
+    /// instructions, this is always 0. For other instructions, this is usually 1.
+    virtual size_t getResultCount() = 0;
+    /// Returns the number of operands consumed by this instruction.
+    virtual size_t getOperandCount() = 0;
+    /// Returns the operand at the specified index. The index must be less than the
+    /// value returned by getOperandCount().
+    virtual Operand getOperand(size_t index) = 0;
+
+    GILValue getResult(size_t index) {
+        assert(index < getResultCount() && "Result index out of range");
+        return GILValue(this, index);
+    }
+    // virtual Type getResultType(size_t index) = 0;
+
+    BasicBlock *getParent() { return parent; }
+    llvm::StringRef getInstName();
+};
```

Comment:
it could be useful to add functions such as `isTerminator()` or whatever

Reply:
that would just be dyn_cast<TerminatorInst>(this)

Reply:
That could be added in #26